server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +345,266 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    /**
+     * iFace which implements the query timeout / cancellation logic
+     */
+    public interface QueryCancellable {
+
+        void checkCancelled();
+    }
+
+    /**
+     * Wraps an {@link IndexReader} with a {@link QueryCancellable}.
+     */
+    static class CancellableDirectoryReader extends FilterDirectoryReader {
+
+        private final Holder<QueryCancellable> cancellableHolder;
+
+        private CancellableDirectoryReader(DirectoryReader in, Holder<QueryCancellable> cancellableHolder) throws IOException {
+            super(in, new SubReaderWrapper() {
+                @Override
+                public LeafReader wrap(LeafReader reader) {
+                    return new CancellableLeafReader(reader, cancellableHolder);
+                }
+            });
+            this.cancellableHolder = cancellableHolder;
+        }
+
+        private void setCancellable(QueryCancellable cancellable) {
+            this.cancellableHolder.set(cancellable);
+        }
+
+        @Override
+        protected DirectoryReader doWrapDirectoryReader(DirectoryReader in) {
+            throw new UnsupportedOperationException("doWrapDirectoryReader() should never be invoked");
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            return in.getReaderCacheHelper();
+        }
+    }
+
+    /**
+     * Wraps a {@link FilterLeafReader} with a {@link QueryCancellable}.
+     */
+    static class CancellableLeafReader extends FilterLeafReader {
+
+        private final Holder<QueryCancellable> cancellable;
+
+        private CancellableLeafReader(LeafReader leafReader, Holder<QueryCancellable> cancellable)  {
+            super(leafReader);
+            this.cancellable = cancellable;
+        }
+
+        @Override
+        public PointValues getPointValues(String field) throws IOException {
+            final PointValues pointValues = in.getPointValues(field);
+            if (pointValues == null) {
+                return null;
+            }
+            return (cancellable.get() != null) ? new ExitablePointValues(pointValues, cancellable.get()) : pointValues;
+        }
+
+        @Override
+        public Terms terms(String field) throws IOException {
+            Terms terms = in.terms(field);
+            if (terms == null) {
+                return null;
+            }
+            return (cancellable.get() != null && terms instanceof CompletionTerms == false) ?
```

Comment:
can you add a comment about why we're not wrapping CompletionTerms? I also wonder that this `instanceof` logic might be fragile as terms may be wrapped e.g. if document-level security is used?

Reply:
Added a comment.

- What other wrapping classes could we encouter?
- Should we have another ExitableTerms class that extends CompletionTerms and also wrap those with cancellation?